server/src/main/java/org/elasticsearch/index/codec/tsdb/es819/ES819TSDBDocValuesProducer.java

```diff
@@ -380,19 +384,29 @@ public long cost() {
 
             @Override
             public BlockLoader.Block tryRead(BlockLoader.BlockFactory factory, BlockLoader.Docs docs, int offset) throws IOException {
-                if (valuesSorted && ords instanceof BaseDenseNumericValues denseOrds) {
-                    int firstDoc = docs.get(offset);
-                    denseOrds.advanceExact(firstDoc);
-                    long startValue = denseOrds.longValue();
-                    final int docCount = docs.count();
-                    int lastDoc = docs.get(docCount - 1);
-                    long lastValue = denseOrds.lookAheadValueAt(lastDoc);
-                    if (lastValue == startValue) {
-                        BytesRef b = lookupOrd(Math.toIntExact(startValue));
-                        return factory.constantBytes(BytesRef.deepCopyOf(b), docCount - offset);
+                if (ords instanceof BaseDenseNumericValues denseOrds) {
+                    if (valuesSorted || entry.termsDictEntry.termsDictSize == 1) {
+                        int firstDoc = docs.get(offset);
+                        denseOrds.advanceExact(firstDoc);
+                        long startValue = denseOrds.longValue();
+                        final int docCount = docs.count();
+                        int lastDoc = docs.get(docCount - 1);
+                        long lastValue = denseOrds.lookAheadValueAt(lastDoc);
+                        if (lastValue == startValue) {
+                            BytesRef b = lookupOrd(Math.toIntExact(startValue));
+                            return factory.constantBytes(BytesRef.deepCopyOf(b), docCount - offset);
+                        }
+                        // TODO: Since ordinals are sorted, start at 0 (offset by startValue), scan until lastValue,
+                        // then fill remaining positions with lastValue.
+                        return null;
+                    }
+                    try (var builder = factory.singletonOrdinalsBuilder(this, docs.count() - offset, true)) {
+                        BlockLoader.SingletonLongBuilder delegate = new SingletonLongToSingletonOrdinalDelegate(builder);
```

Comment:
Any subclasses of `BlockLoader.Block` aren't accessible from server. So instead of consuming the block that denseOrds would produce, I have a delegate builder that translates the long values to int values (given that singleton ordinals works with int[]).